File: bookstore-website/src/types/order.ts
export type OrderStatus =
  | 'PENDING'
  | 'CONFIRMED'
  | 'SHIPPING'
  | 'DELIVERED'
  | 'CANCELLED'

export type PaymentMethod = 'BANK_TRANSFER_QR' | 'COD' 

export type OrderPaymentMethod =
  | PaymentMethod
  | 'CASH'
  | 'BANK_TRANSFER'
  | 'VNPAY'
  | 'MOMO'

export type PaymentStatus = 'PENDING' | 'PAID' | 'FAILED' | 'CANCELLED'

export type OrderPaymentStatus = PaymentStatus | 'UNPAID' | 'REFUNDED'

export type ShippingMethod = 'DELIVERY' | 'PICKUP'

export type OrderItemType = 'PHYSICAL_BOOK' | 'DIGITAL_ASSET'

export type CreateOrderRequest = {
  cartItemIds: string[]
  addressId: string | null
  shippingMethod: ShippingMethod
  paymentMethod: PaymentMethod
  bookCouponCode?: string | null
  shippingCouponCode?: string | null
  note?: string | null
}

export type CreateOrderResponse = {
  orderId: string
  orderCode: string
  paymentMethod: PaymentMethod
  paymentStatus: PaymentStatus
  totalAmount: number
  transferContent: string
}

export type UpdateOrderStatusRequest = {
  status: OrderStatus
}

export type OrderItemResponse = {
  id: string
  itemType: OrderItemType
  bookId: string
  digitalAssetId: string | null
  bookTitle: string
  unitPrice: number
  quantity: number
  lineTotal: number
}

export type OrderResponse = {
  orderId: string
  orderCode: string
  userId: string
  items: OrderItemResponse[]
  productTotal: number
  totalAmount: number
  discountAmount: number
  shippingFee: number
  shippingDiscount: number
  couponDiscount: number
  finalAmount: number
  couponId: string | null
  couponCode: string | null
  bookCouponId: string | null
  bookCouponCode: string | null
  shippingCouponId: string | null
  shippingCouponCode: string | null
  paymentMethod: OrderPaymentMethod
  paymentStatus: OrderPaymentStatus
  status: OrderStatus
  receiverName: string
  receiverPhone: string
  receiverAddress: string
  createdAt: string
  updatedAt: string
  cancelledAt: string | null
}

export type OrderTimelineEventType =
  | 'ORDER_CREATED'
  | 'COUPON_APPLIED'
  | 'PAYMENT_PENDING'
  | 'PAYMENT_PAID'
  | 'ORDER_STATUS_CHANGED'
  | 'SHIPMENT_ASSIGNED'
  | 'SHIPMENT_STATUS_CHANGED'
  | 'ORDER_CANCELLED'
  | 'STOCK_ROLLED_BACK'
  | 'COUPON_ROLLED_BACK'
  | 'RETURN_REQUESTED'
  | 'RETURN_APPROVED'
  | 'RETURN_REJECTED'
  | 'RETURN_CANCELLED'
  | 'REFUND_INTERNAL_APPROVED'
  | 'STOCK_RESTOCKED_FROM_RETURN'

export type OrderTimelineEventResponse = {
  id: string
  orderId: string
  eventType: OrderTimelineEventType | string
  title: string
  description: string | null
  oldStatus: string | null
  newStatus: string | null
  actorName: string | null
  actorRole: string | null
  createdAt: string
  metadata: string | null
}

File: bookstore-mobile/app/src/main/java/com/bookstore/mobile/shared/model/Order.kt
package com.bookstore.mobile.shared.model

data class Order(
    val id: String,
    val orderCode: String,
    val items: List<OrderItem>,
    val productTotal: Double,
    val shippingFee: Double,
    val shippingDiscount: Double,
    val couponDiscount: Double,
    val totalAmount: Double,
    val paymentMethod: String,
    val paymentStatus: String,
    val status: String,
    val receiverName: String,
    val receiverPhone: String,
    val receiverAddress: String,
    val createdAt: String,
    val updatedAt: String,
)

data class OrderItem(
    val id: String,
    val bookId: String,
    val title: String,
    val unitPrice: Double,
    val quantity: Int,
    val lineTotal: Double,
)

data class CheckoutResult(
    val orderId: String,
    val orderCode: String,
    val paymentMethod: String,
    val paymentStatus: String,
    val totalAmount: Double,
    val transferContent: String?,
    val orderStatus: String?,
)

data class BestCouponSuggestion(
    val available: Boolean,
    val couponCode: String?,
    val couponType: String?,
    val discountAmount: Double,
    val finalAmountEstimate: Double,
    val label: String?,
    val reason: String?,
)

data class OrderTimelineEvent(
    val id: String,
    val eventType: String,
    val title: String,
    val description: String?,
    val oldStatus: String?,
    val newStatus: String?,
    val actorName: String?,
    val actorRole: String?,
    val createdAt: String,
)
